scarb/src/core/lockfile.rs

```diff
@@ -0,0 +1,244 @@
+#![allow(dead_code)]
+
+use anyhow::{anyhow, Context, Result};
+use camino::Utf8Path;
+use semver::Version;
+use serde::{Deserialize, Serialize};
+use serde_repr::{Deserialize_repr, Serialize_repr};
+use std::collections::BTreeSet;
+use toml_edit::Document;
+
+use crate::core::{PackageId, PackageName, Resolve, SourceId};
+use crate::internal::fsx;
+
+const HEADER: &str = "# Code generated by scarb DO NOT EDIT.";
+
+#[derive(
+    Default, PartialEq, Eq, Clone, Copy, Debug, PartialOrd, Ord, Serialize_repr, Deserialize_repr,
+)]
+#[repr(u8)]
+pub enum LockVersion {
+    #[default]
+    V1 = 1,
+}
+
+#[derive(Debug, Eq, PartialEq, Default, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct Lockfile {
+    pub version: LockVersion,
+    #[serde(rename = "package")]
+    #[serde(default = "BTreeSet::new")]
+    #[serde(skip_serializing_if = "BTreeSet::is_empty")]
+    pub packages: BTreeSet<PackageLock>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct PackageLock {
+    pub name: PackageName,
+    pub version: Version,
+    #[serde(skip_serializing_if = "skip_path_source_id")]
+    pub source: Option<SourceId>,
+    #[serde(default = "BTreeSet::new")]
+    #[serde(skip_serializing_if = "BTreeSet::is_empty")]
+    pub dependencies: BTreeSet<PackageName>,
+}
+
+fn skip_path_source_id(sid: &Option<SourceId>) -> bool {
+    sid.map(|sid| sid.is_path()).unwrap_or(true)
+}
+
+impl Lockfile {
+    pub fn new(packages: impl Iterator<Item = PackageLock>) -> Self {
+        Self {
+            version: Default::default(),
+            packages: packages.collect(),
+        }
+    }
+
+    pub fn from_resolve(resolve: &Resolve) -> Self {
+        let packages = resolve.package_ids().map(|package| {
+            let deps = resolve
+                .package_dependencies(package)
+                .map(|dep| dep.name.clone());
+            PackageLock::new(&package, deps)
+        });
+        Self::new(packages)
+    }
```

Comment:
We should not include implicit dependencies in the lockfile. Users have no control of their existence.

Topic for separate PR